evalml/tests/pipeline_tests/test_time_series_pipeline.py

```diff
@@ -1724,3 +1724,82 @@ def test_drop_time_index_woodwork(ts_data, time_series_regression_pipeline_class
 
     assert isinstance(X_t.index, pd.DatetimeIndex)
     assert isinstance(y_t.index, pd.DatetimeIndex)
+
+
+@pytest.mark.parametrize(
+    "forecast_horizon,gap,max_delay",
+    [(1, 0, 1), (1, 1, 2), (2, 0, 2), (3, 1, 2), (1, 2, 2), (2, 7, 3), (3, 2, 4)],
+)
+@pytest.mark.parametrize(
+    "freq",
+    ["W", "D", "M", "Y", "Q"],
+)
+@patch("evalml.pipelines.components.RandomForestRegressor.fit")
+@patch("evalml.pipelines.components.RandomForestRegressor.predict")
+def test_dates_needed_for_prediction(
+    mock_predict,
+    mock_fit,
+    forecast_horizon,
+    gap,
+    max_delay,
+    freq,
+    ts_data,
+    time_series_regression_pipeline_class,
+):
+    X, X_t, y = ts_data(freq=freq)
+    X.ww.set_time_index("date")
+
+    pipeline = time_series_regression_pipeline_class(
+        parameters={
+            "pipeline": {
+                "gap": gap,
+                "max_delay": max_delay,
+                "time_index": "date",
+                "forecast_horizon": forecast_horizon,
+            },
+            "Time Series Featurizer": {
+                "gap": gap,
+                "max_delay": max_delay,
+                "time_index": "date",
+                "forecast_horizon": forecast_horizon,
+            },
+        },
+    )
+    pipeline.fit(X, y)
+
+    prediction_date = pd.Timestamp("2022-01-31")
+    beginning_date, end_date = pipeline.dates_needed_for_prediction(prediction_date)
+    assert beginning_date <= end_date
+    assert end_date < prediction_date
+    date_diff = pipeline.forecast_horizon + pipeline.max_delay + pipeline.gap
+    assert end_date == beginning_date + pd.tseries.frequencies.to_offset(
+        f"{date_diff}{pipeline.frequency}",
+    )
```

Comment:
Do we want to reparametrize this so that the `pytest` parameters contain the answers to the incoming inputs and just test them against each other?

Reply:
I think it'll get pretty messy including the answers with the different combinations of parameters and frequencies above. I could do so but I would need to narrow down the scope of combinations.